cpp/pagerank_module/algorithm/pagerank.cpp

```diff
@@ -0,0 +1,140 @@
+  std::vector<double> new_rank(graph.GetNodeCount(), 0);
+  // Calculate sums of PR(page)/C(page) scores for the entire block (from lo to
+  // hi edges).
+  for (std::size_t edge_id = lo; edge_id < hi; edge_id++) {
+    const auto [source, target] = graph.GetOrderedEdges()[edge_id];
+    // Add the score of target node to the sum.
+    new_rank[source] += old_rank[target] / graph.GetOutDegree(target);
+  }
+  new_rank_promise.set_value(new_rank);
+}
+
+void AddCurrentBlockToRankNext(const PageRankGraph &graph, const double damping_factor,
+                               const std::vector<double> &block, std::vector<double> &rank_next) {
+  // The block vector contains partially precalculated sums of PR(page)/C(page)
+  // for each node. Node index in the block vector corresponds to the node index
+  // in the rank_next vector.
+  for (std::size_t node_index = 0; node_index < block.size(); node_index++) {
+    rank_next[node_index] += damping_factor * block[node_index];
+  }
+}
+
+void CompleteRankNext(const PageRankGraph &graph, const double damping_factor, std::vector<double> &rank_next) {
+  while (rank_next.size() < graph.GetNodeCount()) {
+    rank_next.push_back((1.0 - damping_factor) / graph.GetNodeCount());
+  }
+}
+
+bool CheckContinueIterate(const std::vector<double> &rank, const std::vector<double> &rank_next,
+                          const std::size_t max_iterations, const double stop_epsilon,
+                          const std::size_t number_of_iterations) {
+  if (number_of_iterations == max_iterations) {
+    return false;
+  }
+  for (std::size_t node_id = 0; node_id < rank.size(); node_id++) {
+    if (std::abs(rank[node_id] - rank_next[node_id]) > stop_epsilon) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void NormalizeRank(std::vector<double> *rank) {
+  const double sum = std::accumulate(rank->begin(), rank->end(), 0.0);
+  for (double &value : *rank) {
+    value /= sum;
+  }
+}
+
+std::vector<double> ParallelIterativePageRank(const PageRankGraph &graph, std::size_t max_iterations,
+                                              double damping_factor, double stop_epsilon) {
+  const std::uint64_t number_of_threads = std::thread::hardware_concurrency();
+
+  std::vector<std::uint64_t> borders;
+  CalculateOptimalBorders(graph, number_of_threads, &borders);
+
+  std::vector<double> rank(graph.GetNodeCount(), 1.0 / graph.GetNodeCount());
+  // Because we increment number_of_iterations at the end of while loop.
+  bool continue_iterate = max_iterations != 0;
+
+  std::size_t number_of_iterations = 0;
+  while (continue_iterate) {
+    std::vector<std::promise<std::vector<double>>> page_rank_promise(number_of_threads);
+    std::vector<std::future<std::vector<double>>> page_rank_future(number_of_threads);
+    std::transform(page_rank_promise.begin(), page_rank_promise.end(), page_rank_future.begin(),
+                   [](auto &pr_promise) { return pr_promise.get_future(); });
```

Comment:
you mean to call `reserve` here on `page_rank_future` and use `std::back_inserter(page_rank_future)` in the `std::transform`.

Reply:
Why reserve when the container's size is initialized with `number_of_threads`?

Reply:
Yep, exactly!